### PR TITLE
Interpreter design pattern

### LINKED ISSUE #210: Interpreter
I'm working on it.

### PR DESCRIPTION
**Pull request by fade2black**
closes #210

### CODE REVIEW

**Comment by pickfire** [comment]
@fade2black Can you please mark this as draft pull request? I don't see any difference.

**Comment by simonsan** [comment]
> Can you please mark this as draft pull request? I don't see any difference.

Good catch, done that now :)

**Comment by fade2black** [comment]
I could add more on `macro_rules!`. For example creating DSLs as Interpreter pattern in the Discussion section. But I am not sure whether or not it'll make the chapter too long.

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:6]:
```suggestion
then the problem instances might be expressed in a simple
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:5]:
```suggestion
If a problem occurs very often and requires long and repetitive steps to solve it,
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:8]:
```suggestion
interpreting the sentences written in this simple languge.
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:15]:
```suggestion
Imagine that our work is translating simple mathematical expressions into
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:19]:
```suggestion
four operations `+, -, /, *` and a pair of parenthesis `(, )`.
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:20]:
```suggestion
For example, expression `2 + 4` could be translated into
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:30]:
```suggestion
provide the Interpreter with an expressions and get Assembly
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:38]:
```suggestion
but it at least computes the expression correctly.
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:43]:
```suggestion
The grammar for a set of expressions over `0,...,9, +,-,*,/,(,)` is
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:57]:
```suggestion
a recursive descent parser. The following code
```
First time I heard to use recursive descent for parsing, usually I heard it is used for machine learning.

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:60]:
```suggestion
(unbalanced parentheses or missing digit/operator for example).
```

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md]:
```suggestion
    let a = 3;
```

**Comment by simonsan** [comment]
Is this ready to review? ;)

**Comment by fade2black** [comment]
@simonsan of course. I'll make changes as I get feedbacks.

**Comment by simonsan** [comment]
> @simonsan of course. I'll make changes as I get feedbacks.

Ok ;-) wasn't sure if it's still WIP hence I thought I'd ask. Will take a look into it today/tomorrow. 👍🏽

**Comment by fade2black** [comment]
@simonsan decided not go far without people's feedaback :-)

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:131]: Why not `push_str`?

Comment (fade2black) [reply patterns/interpreter.md:131]: @pickfire because I store each instruction as a separate entry in a vector.

Comment (fade2black) [reply patterns/interpreter.md:131]: For example, I might need to do additional scan over instructions, or simple compute number of instructions,..etc.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:7]:
```suggestion
language and an interpreter object could solve it by
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:30]:
```suggestion
provide the Interpreter with an expression and get Assembly
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:40]: This is confusing, I had to read it multiple times. Couldn't we just delete it? I don't think it's necessary. It's pretty clear that this is an example and we are not building a modern compiler in my opinion :stuck_out_tongue: 
What do you think?

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:53]:
```suggestion
factor -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
```

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:40]: sure

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:39]: Should we give a brief explanation about what a grammar is? And explain this example in particular?
For example why both `exp` and `term` are necessary.
Otherwise this chapter is readable only by people with a MSc degree in computer science 😅

Comment (fade2black) [reply patterns/interpreter.md:39]: @MarcoIeni  Sure, I can explain, but I think we shouldn't go beyond simply mentioning what a CFG is and that `expr`, `terms...` and `factor` are the result of grammar transformation. In any case strong knowledge of CS is still needed :-D . I don't mind detailed explanation of CFG and related topics, but to what extend?

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:41]:
```suggestion
(CFG) defines a set of rules that describe all possible strings
```
Do you like this version better?

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:43]:
```suggestion
These strings usually have some well defined structure.
```
Also, is `some` correct here?

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:44]:
```suggestion
For example, strings of all balanced parentheses:
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:47]:
```suggestion
(), (()), ()(), ()(()), ..
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:50]: I count four items here :stuck_out_tongue:

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:56]:
```suggestion
symbol `S` and production rules as
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:61]:
```suggestion
S -> (S)
S -> ()
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:71]: Here we have to start a new section, just to make it clear that the previous example is finished.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md]: Thank you for providing the small explanation, with an example, too!
I think it works great :)

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:38]:
```suggestion

In order to understand the Interpreter pattern you need to know a little bit
about formal language theory.
If you already know what a grammar is feel free to skip this section,
otherwise fasten your seat belts.

```
Do you like this?

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:104]:
```suggestion
Furthermore, the code panics when an expression is syntactically wrong
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:108]: first question that came to my mind: "what's cur?"

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:197]: you are not using cur after this, so there is no need to take it as argument.
It could be created inside `single_expr`

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:71]: Actually, I think we should move this section, because it is in the middle of the assembly stuff. I would put this before the assembly discussion. What do you think?

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:113]: It stands for current? Maybe we should call this `index` or something like that?
In general I like extended nouns, but this might be only my personal preference.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:181]: I agree with fade2black, it is better to store each instruction in a separate entry of the vec.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:117]: What if here we return `None` instead of the special character `$`?
So the return type of this function is an `Option`.

However, maybe the best thing would be to implement an Iterator.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:113]: > first question that came to my mind: "what's cur?"
`cur` is the index pointing to the current symbol.
Yeah, we could rename it for readability.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:202]: @MarcoIeni  In fact I think the best solution is to wrap everything into a struct `Parser`, define `index` and other necessary vars like `index`, and hence we could git rid of passing around `index`. I didn't t want to bloat the code.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:117]: Is `$` too confusing? Adding `Option` will chain other changes. Lines may increase. Is it ok to add more lines? Last time I wrote such code I got feedback that "too many lines".

**Comment by pickfire** [comment]
I personally think the example is overly complicated.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:76]: @MarcoIeni I don't mind. Could you update yourself?

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md]: @MarcoIeni fixed. :-)

**Comment by fade2black** [comment]
@pickfire I agree, the assembly language may seem daunting. I could remove this part. How about simply converting normal (infix) expressions into postfix or prefix expressions? For example, `2 + 3` => `2 3 +` or `+ 2 3`. I took this example from GoF. [Wikipedia](https://en.wikipedia.org/wiki/Interpreter_pattern) also explains by giving the same examples, but much more simpler. Only two operations `+` and `-`.

**Comment by simonsan** [comment]
> 
> 
> @pickfire @MarcoIeni @simonsan I agree, the assembly language may seem to be daunting. I could remove this part. How about simply converting normal (infix) expressions into postfix or prefix expressions? For example, `2 + 3` => `2 3 +` or `+ 2 3`. I took this example from GoF. [Wikipedia](https://en.wikipedia.org/wiki/Interpreter_pattern) also explains by giving the same examples, but much more simpler. Only two operations `+` and `-`.

I think the easier the example, the more broken down to the basics the better for Rustlings the more viable for this repository it is. So for sure, if you think it's easier for people reading your article, a topic change might be viable. ;)

**Comment by fade2black** [comment]
@simonsan Ok, let me simply it and I'll let you know to review. 🙂

**Comment by simonsan** [comment]
> @simonsan Ok, let me simply it and I'll let you know to review. 🙂

Ok :)

**Comment by pickfire** [comment]
That seemed better. I have tons of articles in this-week-in-rust.org talking about parsing, some more with different methods. Maybe we can just point them to those articles, those are very detailed step by step guide using specific algorithms.

**Comment by simonsan** [comment]
> That seemed better. I have tons of articles in this-week-in-rust.org talking about parsing, some more with different methods. Maybe we can just point them to those articles, those are very detailed step by step guide using specific algorithms.

You could post them here if they are viable so we can add them to the `see also` section. But external articles are not really a replacement for examples in our articles, so not sure how much that could make the examples less complicated or something. Or did I understand you wrong?

**Comment by fade2black** [comment]
@simonsan Ready. You can review.

File: patterns/interpreter.md
Comment (simonsan) [inline patterns/interpreter.md:138]: For me it's ending a bit abruptly. We could mak a small discussion point here or something?

Comment (fade2black) [reply patterns/interpreter.md:138]: @simonsan  Yeah, I'm going to add. It is not finished. I want to also mention about `macro_rules` how to create DSLs
I mentioned you simply to get feedback on simplicity of the example.

Comment (simonsan) [reply patterns/interpreter.md:139]: Ah ok I get it, yeah, I think that it is much simpler and a good take. 👍🏽

**Comment by fade2black** [comment]
@simonsan @pickfire @MarcoIeni If you have anything to add plz let me know.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:11]:
```suggestion
simple language.
Basically, for any kind of problems we define:

- a [domain specific language](https://en.wikipedia.org/wiki/Domain-specific_language),
- a grammar for this language
- an interpreter that solves the problem instances.
```

This sentence is not that clear. Check if you like something like that or rephrase your version a bit.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:15]:
```suggestion
Our goal is to translate simple mathematical expressions into
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:18]:
```suggestion
For simplicity, our expressions consist of ten digits `0`, ..., `9` and two
operations `+`, `-`. For example, the expression `2 + 4` is translated into `2 4  +`.
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:58]:
```suggestion
define a context free grammar for a set of infix expressions over `0`, ..., `9`, `+`, and `-`,
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:62]:
```suggestion
- terminal symbols: `0`, ..., `9`, `+`, `-`
- non-terminal symbols: `exp`, `term`
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:74]:
```suggestion
to do with it. For example, we might need to remove left recursion.
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:75]:
```suggestion
For more details please see [Compilers: Principles, Techniques, and Tools](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:73]: I don't understand how "in fact" is connected with the rest.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:82]:
```suggestion
We simply implement a recursive descent parser.
For simplicity's sake, the code panics when an expression
is syntactically wrong (for example `2-34` or `2+5-` are wrong according to the
grammar definition).
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:91]:
```suggestion
    pub fn new(infix: &'a str) -> Self {
```

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:75]: @MarcoIeni This CFG is not suitable yet to parse with a recursive descent parser. It has left recursion which should be eliminated. But I didn't go into details.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:100]: what's the point of having this function? 🤔
It's the same as `exp`

Comment (fade2black) [reply patterns/interpreter.md:100]: @MarcoIeni  I just follow the convention. Trait or struct must implement the function `interpret `, though I don't know why it is required. Look [https://en.wikipedia.org/wiki/Interpreter_pattern](url) or GoF book.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:154]:
```suggestion
Rust language has `macro_rules!` that allow to define special
```

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:155]: I think that "how to" is not correct here.

File: patterns/interpreter.md
Comment (MarcoIeni) [inline patterns/interpreter.md:157]: is this example part of the Interpreter design pattern? Or it's just an example in which you explain macro rules?

**Comment by MarcoIeni** [comment]
The new example was great! Very easy to understand, thank you!

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:157]: @MarcoIeni Using `macro_rules!` is one way to implement interpreter pattern. I am not explaining here how to use `macro_rules!`. The basic idea is that you express a problem in a special syntax like `norm!(...)` rather `let v = vec![...]; norm(&a);...`.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:155]: @MarcoIeni Why "how to" is not correct? This is from Rust by Examples book: 
> instead of generating a function call, macros are expanded into source code that gets compiled with the rest of the program

But how, I mean according to what rules, it will be expanded depends on us, e.g., on how we define the rules.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md]: @MarcoIeni Then we could remove unnecessary `String` instances from `main`. I mean simply call `intr.interpret("2+3")`.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:77]: Ok, let's see if other reviewers find that there is a problem here.
If nobody replies within a week or so we can mark this as resolved.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:153]: no, sorry, I mean from an English grammar point of view.
Maybe it is better:

```suggestion
syntax and rules on how to expand this syntax into source code.
```

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:155]: Ok.
In the previous version of this article you used macro rules to rewrite the same example you first wrote in code.
This was really effective in my opinion because you clearly showed the power of macro rules.
I wonder if you think this would be useful also here, instead of doing a different example.
I would like to hear also other reviewer's opinion.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:104]: Ok, if you like it you could delete this function and rename `exp` into `interpret` (and make it pub).
Otherwise if you like things as they are now just mark this conversation as resolved :)

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:77]: @MarcoIeni We could remove this part if it is confusing. I put it here just in case someone asks "how did they come up with this Rust code from the gramma?". In fact implementation is up to implementer. I'd personally prefer manually coding if the grammar is "simple" enough to be addressed with a recursive descent parser. Another one could prefer tools such as Yacc, Bison.... or just use table. You know this is outside of our scope. All of them could be found in a book on compilers or specifically parsers. As I mentioned, this pattern is NOT at all about parsers. So, if you feel this "in fact" part does not fit here, let's remove. Or you could suggest another wording.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:155]: @MarcoIeni Ok, let's see what others think. I could rewrite the example if we come to a consensus.

Comment (simonsan) [reply patterns/interpreter.md:148]: For me either way is nice, if you both think it's nice to use the same "problem" to show the power of macros, why not. :)

File: patterns/interpreter.md
Comment (simonsan) [reply patterns/interpreter.md]: I wrapped the text block under it into a "Note", so it's more seen as some additional explanation. Should be fine imho.

File: patterns/interpreter.md
Comment (MarcoIeni) [reply patterns/interpreter.md:148]: it's up to you fade2black :)

File: patterns/interpreter.md
Comment (pickfire) [inline patterns/interpreter.md:57]: I am kinda confused at this part other than the other parts, maybe my CS knowledge isn't that good yet. Maybe I get a little bit but still not quite understand.

Comment (fade2black) [reply patterns/interpreter.md:57]: @pickfire which part is confusing? Is explanation too short/not clear? Or is CFGs is not necessary here?

Comment (pickfire) [reply patterns/interpreter.md:57]: I mean I find it harder to understand, maybe not very clear or because the prerequisite knowledge required is too high, not sure if it is necessary.

Comment (fade2black) [reply patterns/interpreter.md:57]: @pickfire 
> the prerequisite knowledge required is too high

In this section I very briefly describe [Context Free Grammars](https://en.wikipedia.org/wiki/Context-free_grammar). This topic is typically covered in Formal languages or Compilers course, and it is typically given at senior undergrad or graduate CS/CE level. I agree that for a someone who never saw CFG this section isn't enough to convey the whole picture about the subject. We could add a note that it is not essential and just simply a reminder and you can skip it.
 
Next, the reason why I am mentioning, here, CFGs is because the Interpreter pattern is closely related to formal languages, though it is not about writing parsers. Initially I didn't include this section. Later I included it on reviewers' recommendations. 

We could remove this section. What do @simonsan and @MarcoIeni think?

File: patterns/interpreter.md
Comment (simonsan) [inline patterns/interpreter.md:30]:
```suggestion
Basically, a [Context Free Grammar (CFG)](https://en.wikipedia.org/wiki/Context-free_grammar)
```

File: patterns/interpreter.md
Comment (simonsan) [reply patterns/interpreter.md:57]: Hmm, I know what @pickfire means, but I would actually keep it as is and wait for feedback of a broader mass of people. If there are many questions coming up, we can refactor in the future, answer frequently asked questions etc. 

> We could add a note that it is not essential and just simply a reminder and you can skip it.

Indeed that is a viable approach imho 👍🏽 

Y

Comment (fade2black) [reply patterns/interpreter.md]: Removed CFG section.

File: patterns/interpreter.md
Comment (fade2black) [reply patterns/interpreter.md:113]: I just left as is.

**Review by pickfire — APPROVED** [review]
Looks good to me.

**Review by simonsan — APPROVED** [review]

**Comment by simonsan** [comment]
Thank you for all your contributions @fade2black 👍🏽 <3

**Pull request merged**
